<!DOCTYPE html>
<html lang="en-us">
  <head>
    <meta charset="UTF-8">
    <title>Altre by Mathelab</title>
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link rel="stylesheet" type="text/css" href="stylesheets/normalize.css" media="screen">
    <link href='https://fonts.googleapis.com/css?family=Open+Sans:400,700' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" type="text/css" href="stylesheets/stylesheet.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/github-light.css" media="screen">
  </head>
  <body>
    <section class="page-header">
      <h1 class="project-name">Altre</h1>
      <h2 class="project-tagline">ALTered Regulatory Elements</h2>
      <a href="https://github.com/Mathelab/ALTRE" class="btn">View on GitHub</a>
      <!-- <a href="https://github.com/Mathelab/ALTRE/zipball/master" class="btn">Download .zip</a> --> 
      <a href="https://github.com/Mathelab/ALTRE/tarball/master" class="btn">Download .tar.gz</a>
      <a href="https://mathelab.github.io/ALTRE/vignette.html" class="btn"> View Vignette (HTML)</a>
      <a href="https://mathelab.github.io/ALTRE/vignette.pdf" class="btn"> View Vignette (PDF)</a>
    </section>

    <section class="main-content">
      <h4>
<a id="install-from-github" class="anchor" href="#install-from-github" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Install From Github</h4>

<pre><code>install.packages("devtools")
</code></pre>

<h4>
<a id="install-bioconductor-packages" class="anchor" href="#install-bioconductor-packages" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>install Bioconductor packages</h4>

<pre><code>source("http://bioconductor.org/biocLite.R")
BiocInstaller::biocLite(c('GenomeInfoDb',
                        'IRanges',
                        'DESeq2',
                        'GenomicAlignments',
                        'SummarizedExperiment',
                        'GenomicRanges',
                        'Rsamtools',
                        'org.Hs.eg.db',
                        'clusterProfiler',
                        'ensembldb', 
                        'EnsDb.Hsapiens.v75',
                        'GO.db'))
</code></pre>

<h1>
<a id="install-the-package" class="anchor" href="#install-the-package" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>install the package</h1>

<pre><code>devtools::install_github("mathelab/ALTRE")
</code></pre>

<p>When installing on Linux, if you get an rJava package installation error. Please run the following two lines in the console:</p>

<pre><code>sudo apt-get install openjdk-7-*
sudo R CMD javareconf
</code></pre>

<p>On Windows, if you get an installation error then first run the following lines of code in as well:</p>

<pre><code>install.packages(c("htmltools","httpuv","evaluate","markdown"))
</code></pre>

<p>When installing on OSC cluster, if you get an installation error then first run the following lines of code in your console prior to trying to install:</p>

<pre><code>sudo R CMD javareconf -e
export JAVA_CPPFLAGS='-O2'
module load gnu/5.2.0
module load R/3.3.1
</code></pre>

<h4>
<a id="installation-walk-through-animation" class="anchor" href="#installation-walk-through-animation" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Installation Walk-through Animation</h4>

<p><img src="https://raw.githubusercontent.com/Mathelab/ALTRE/master/inst/img/ALTREinstall.gif" alt=""></p>

<h3>
<a id="to-run" class="anchor" href="#to-run" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>To Run</h3>

<p>To launch the Shiny app inside R, run</p>

<pre><code>library(ALTRE)
runShinyApp()
</code></pre>

<h4>
<a id="shiny-app-preview" class="anchor" href="#shiny-app-preview" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Shiny App Preview</h4>

<p><img src="https://raw.githubusercontent.com/Mathelab/ALTRE/master/inst/img/ALTRErun.gif" alt=""></p>

<h2>
<a id="vignette" class="anchor" href="#vignette" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Vignette</h2>

<p>A vignette (which provides an overview of the package via step-by-step guide through an example dataset) is here:
<a href="https://mathelab.github.io/ALTRE/vignette.html">https://mathelab.github.io/ALTRE/vignette.html</a></p>

<h2>
<a id="data" class="anchor" href="#data" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Data</h2>

<p>A restricted subset of the data with one chromosome (i.e. chromosome 21) can be found on this <a href="http://mathelab.github.io/ALTREsampledata/">page</a>. The corresponding CSV file for input into ALTRE can be downloaded <a href="https://raw.githubusercontent.com/mathelab/ALTREsampledata/master/DNaseEncodeExample.csv">here</a>. Be sure to modify the datapath column of the CSV file so that the appropriate full path of the data files on your local machine is included.</p>

<p>To download the entire data, please use a file download manager to download the files from the links listed below. </p>

<h3>
<a id="alignment-in-bam-format-files" class="anchor" href="#alignment-in-bam-format-files" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Alignment (in BAM format) files:</h3>

<h4>
<a id="a549" class="anchor" href="#a549" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><em>A549</em>:</h4>

<p><a href="https://www.encodeproject.org/files/ENCFF001CLE/@@download/ENCFF001CLE.bam">https://www.encodeproject.org/files/ENCFF001CLE/@@download/ENCFF001CLE.bam</a></p>

<p><a href="https://www.encodeproject.org/files/ENCFF001CLJ/@@download/ENCFF001CLJ.bam">https://www.encodeproject.org/files/ENCFF001CLJ/@@download/ENCFF001CLJ.bam</a></p>

<h4>
<a id="saec" class="anchor" href="#saec" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><em>SAEC</em>:</h4>

<p><a href="https://www.encodeproject.org/files/ENCFF001EFI/@@download/ENCFF001EFI.bam">https://www.encodeproject.org/files/ENCFF001EFI/@@download/ENCFF001EFI.bam</a></p>

<p><a href="https://www.encodeproject.org/files/ENCFF001EFN/@@download/ENCFF001EFN.bam">https://www.encodeproject.org/files/ENCFF001EFN/@@download/ENCFF001EFN.bam</a></p>

<h3>
<a id="peakhotspot-in-bed-format-files" class="anchor" href="#peakhotspot-in-bed-format-files" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Peak/hotspot (in BED format) files:</h3>

<h4>
<a id="a549-1" class="anchor" href="#a549-1" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><em>A549</em>:</h4>

<p><a href="http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeUwDnase/wgEncodeUwDnaseA549HotspotsRep1.broadPeak.gz">http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeUwDnase/wgEncodeUwDnaseA549HotspotsRep1.broadPeak.gz</a></p>

<p><a href="http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeUwDnase/wgEncodeUwDnaseA549HotspotsRep2.broadPeak.gz">http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeUwDnase/wgEncodeUwDnaseA549HotspotsRep2.broadPeak.gz</a></p>

<h4>
<a id="saec-1" class="anchor" href="#saec-1" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><em>SAEC</em>:</h4>

<p><a href="http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeUwDnase/wgEncodeUwDnaseSaecHotspotsRep1.broadPeak.gz">http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeUwDnase/wgEncodeUwDnaseSaecHotspotsRep1.broadPeak.gz</a></p>

<p><a href="http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeUwDnase/wgEncodeUwDnaseSaecHotspotsRep2.broadPeak.gz">http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeUwDnase/wgEncodeUwDnaseSaecHotspotsRep2.broadPeak.gz</a></p>

<h3>
<a id="blacklisted-genomic-regions-for-functional-genomics-analysis" class="anchor" href="#blacklisted-genomic-regions-for-functional-genomics-analysis" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Blacklisted genomic regions for functional genomics analysis</h3>

<p><a href="https://sites.google.com/site/anshulkundaje/projects/blacklists">https://sites.google.com/site/anshulkundaje/projects/blacklists</a></p>

      <footer class="site-footer">
        <span class="site-footer-owner"><a href="https://github.com/Mathelab/ALTRE">Altre</a> is maintained by <a href="https://github.com/Mathelab">Mathelab</a>.</span>

        <span class="site-footer-credits">This page was generated by <a href="https://pages.github.com">GitHub Pages</a> using the <a href="https://github.com/jasonlong/cayman-theme">Cayman theme</a> by <a href="https://twitter.com/jasonlong">Jason Long</a>.</span>
      </footer>

    </section>

  
  </body>
</html>
